File: 2025/03 March/11 Ways to Reach the n'th Stair/main.py
# Link: https://www.geeksforgeeks.org/problems/count-ways-to-reach-the-nth-stair-1587115620/1


class Solution:
    def countWaysRec(self, n, memo):
        if n == 0 or n == 1:
            return 1

        if memo[n] != -1:
            return memo[n]

        memo[n] = self.countWaysRec(n - 1, memo) + self.countWaysRec(n - 2, memo)
        return memo[n]

    def countWays(self, n):
        memo = [-1] * (n + 1)
        return self.countWaysRec(n, memo)


if __name__ == "__main__":
    obj = Solution()
    n = 1
    print(obj.countWays(n))

    n = 2
    print(obj.countWays(n))

    n = 4
    print(obj.countWays(n))

File: 2025/03 March/11 Ways to Reach the n'th Stair/main.cpp
// Link: https://www.geeksforgeeks.org/problems/count-ways-to-reach-the-nth-stair-1587115620/1

#include <iostream>
#include <vector>
using namespace std;

class Solution
{
public:
    int countWaysRec(int n, vector<int> &memo)
    {
        if (n == 0 || n == 1)
            return 1;

        if (memo[n] != -1)
            return memo[n];

        return memo[n] = countWaysRec(n - 1, memo) +
                         countWaysRec(n - 2, memo);
    }

    int countWays(int n)
    {
        vector<int> memo(n + 1, -1);
        return countWaysRec(n, memo);
    }
};

int main()
{
    Solution obj;
    int n = 1;
    cout << obj.countWays(n) << endl;

    n = 2;
    cout << obj.countWays(n) << endl;

    n = 4;
    cout << obj.countWays(n) << endl;
    return 0;
}